modules/tribes/tests/client/components/TribesPage.component.test.js

```diff
@@ -0,0 +1,177 @@
+/* tests to write
+
+when clicking join and not logged in, redirect to tribe page
+when clicking join and logged in, send a request to join to API
+when clicking join and logged in, update the number of tribe members
+when clicking leave, send a request to leave to API
+when clicking leave and logged in, update the number of tribe members
+
+
+*/
+
+import React from 'react';
+import * as rtl from '@testing-library/react';
+import { render, fireEvent, wait } from '@testing-library/react';
+import '@testing-library/jest-dom/extend-expect';
+import '@/config/client/i18n';
+
+import TribesPage from '@/modules/tribes/client/components/TribesPage.component';
+import * as api from '@/modules/tribes/client/api/tribes.api';
+
+jest.mock('@/modules/tribes/client/api/tribes.api');
+
+const dummyTribes = [{
+  _id: 'aaaa',
+  id: 'aaaa',
+  count: 20000,
+  slug: 'hitchhikers',
+  label: 'Hitchhikers',
+}, {
+  _id: 'bbbb',
+  id: 'bbbb',
+  count: 5000,
+  slug: 'snails',
+  label: 'Snails',
+}, {
+  _id: 'cccc',
+  id: 'cccc',
+  count: 1500,
+  slug: 'teddybears',
+  label: 'Teddy Bears',
+}];
+
+const dummyUser = {
+  _id: 'user0',
+  id: 'user0',
+  memberIds: ['cccc', 'aaaa'],
+};
+
+api.read.mockImplementation(async () => dummyTribes);
+api.join.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+api.leave.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+
+
+describe('TribesPage', () => {
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
+
+  describe('not signed in', () => {
+
+    it('fetch tribes from api and show tribes on page', async () => {
+      // first no api should be called
+      expect(api.read).toHaveBeenCalledTimes(0);
+
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // the tribes from api should be displayed
+      const tribes = getAllByRole('listitem');
+      expect(tribes).toHaveLength(dummyTribes.length + 1); // there is additionally the Suggest Tribe
+
+      tribes.slice(0, -1).forEach((tribe, i) => {
+        expect(tribe).toHaveTextContent(dummyTribes[i].label);
+      });
+
+      // during the test the api should be called only once
+      expect(api.read).toHaveBeenCalledTimes(1);
+    });
+
+    it('the join button should be a link to tribe page', async () => {
+      const { getAllByRole } = render(<TribesPage onMembershipUpdated={() => {}} />);
+
+      await wait();
+
+      // join button
+      const tribes = getAllByRole('listitem').slice(0, -1);
```

Comment:
I wonder if it's better to get everything by the button role instead of the list items? Not sure.

Reply:
When the user is signed out, the buttons are just links, so `getAllByRole(button)` won't find anything. However we can do `getAllByText('Join', { selector: 'a' })` instead. The selector part is optional.

Reply:
98c404329bf72eef2a6276eb12484bcc098660f3

Reply:
I think it's fine if you prefer to add some `data-identifier` to those elements (or `id` if you want some specific one) for the tests. Up to you!